### PR TITLE
Reproduce a 'bad hunk' format error

### LINKED ISSUE #165: bad hunk format
Log:

```
{"app":"lookout","level":"info","msg":"fetching references for repository git://github.com/src-d/lookout.git: [refs/heads/master:refs/heads/master refs/pull/164/head:refs/pull/164/head]","time":"2018-08-22T16:54:58.517536018Z"}
{"app":"lookout","duration":253827451,"grpc.code":0,"grpc.method":"GetChanges","grpc.service":"pb.Data","grpc.start_time":"2018-08-22T16:54:58Z","level":"info","msg":"streaming server call finished","span.kind":"server","system":"grpc","time":"2018-08-22T16:54:58.770983492Z"}
{"app":"lookout","duration":256395502,"grpc.code":0,"grpc.method":"NotifyReviewEvent","grpc.service":"pb.Analyzer","grpc.start_time":"2018-08-22T16:54:58Z","level":"info","msg":"streaming client call finished","span.kind":"client","system":"grpc","time":"2018-08-22T16:54:58.772094572Z"}
{"app":"lookout","comments":1,"event-id":"8e7525f1c831ee49115fd843ed6340f086a56009","event-type":2,"head":"refs/pull/164/head","level":"info","msg":"posting analysis","provider":"github","repository":"git://github.com/src-d/lookout.git","time":"2018-08-22T16:54:58.808879789Z"}
{"app":"lookout","error":"bad hunk format","hunk":"@@ -6,20 +6,21 @@ import (\n \t\"database/sql\"\n \t\"fmt\"\n \t\"io\"\n-\t\"io/ioutil\"\n \t\"os\"\n \t\"os/exec\"\n+\t\"strings\"\n \t\"time\"\n \n \t_ \"github.com/lib/pq\"\n+\t\"github.com/stretchr/testify/require\"\n )\n \n // CmdTimeout defines timeout for a command\n var CmdTimeout = time.Minute\n \n // default path to binaries\n-var dummyBin = \"build/bin/dummy\"\n-var lookoutBin = \"build/bin/lookout\"\n+var dummyBin = \"../../build/bin/dummy\"\n+var lookoutBin = \"../../build/bin/lookout\"\n \n // function to stop running commands\n // redefined in StoppableCtx\n@@ -50,7 +51,7 @@ func StoppableCtx() (context.Context, func()) {\n }\n \n // StartDummy starts dummy analyzer with context and optional arguments\n-func StartDummy(ctx context.Context, args ...string) io.Reader {\n+func StartDummy(ctx context.Context, require *require.Assertions, args ...string) io.Reader {\n \tr, outputWriter := io.Pipe()\n \tbuf := \u0026bytes.Buffer{}\n \ttee := io.TeeReader(r, buf)\n@@ -61,31 +62,25 @@ func StartDummy(ctx context.Context, args ...string) io.Reader {\n \tcmd.Stdout = outputWriter\n \tcmd.Stderr = outputWriter\n \terr := cmd.Start()\n-\tif err != nil {\n-\t\tioutil.ReadAll(tee)\n-\t\tfmt.Println(\"can't start analyzer:\")\n-\t\tfmt.Println(err)\n-\t\tfmt.Printf(\"output:\\n %s\", buf.String())\n-\t\tos.Exit(1)\n-\t} else {\n-\t\tgo func() {\n-\t\t\tif err := cmd.Wait(); err != nil {\n-\t\t\t\t// don't print error if analyzer was killed by cancel\n-\t\t\t\tif ctx.Err() != context.Canceled {\n-\t\t\t\t\tioutil.ReadAll(tee)\n-\t\t\t\t\tfmt.Println(\"analyzer exited with error:\", err)\n-\t\t\t\t\tfmt.Printf(\"output:\\n%s\", buf.String())\n-\t\t\t\t\tfailExit()\n-\t\t\t\t}\n+\trequire.NoError(err, \"can't start analyzer\")\n+\n+\tgo func() {\n+\t\tif err := cmd.Wait(); err != nil {\n+\t\t\t// don't print error if analyzer was killed by cancel\n+\t\t\tif ctx.Err() != context.Canceled {\n+\t\t\t\tfmt.Println(\"analyzer exited with error:\", err)\n+\t\t\t\tfmt.Printf(\"output:\\n%s\", buf.String())\n+\t\t\t\t// T.Fail cannot be called from a goroutine\n+\t\t\t\tfailExit()\n \t\t\t}\n-\t\t}()\n-\t}\n+\t\t}\n+\t}()\n \n \treturn tee\n }\n \n // StartServe starts lookout server with context and optional arguments\n-func StartServe(ctx context.Context, args ...string) (io.Reader, io.WriteCloser) {\n+func StartServe(ctx context.Context, require *require.Assertions, args ...string) (io.Reader, io.WriteCloser) {\n \tr, outputWriter := io.Pipe()\n \tbuf := \u0026bytes.Buffer{}\n \ttee := io.TeeReader(r, buf)\n@@ -97,85 +92,59 @@ func StartServe(ctx context.Context, args ...string) (io.Reader, io.WriteCloser)\n \tcmd.Stderr = outputWriter\n \n \tw, err := cmd.StdinPipe()\n-\tif err != nil {\n-\t\tfmt.Println(\"can't start server:\")\n-\t\tfmt.Println(err)\n-\t\tos.Exit(1)\n-\t}\n+\trequire.NoError(err, \"can't start server\")\n \n \terr = cmd.Start()\n-\tif err != nil {\n-\t\tioutil.ReadAll(tee)\n-\t\tfmt.Println(\"can't start server:\")\n-\t\tfmt.Println(err)\n-\t\tfmt.Printf(\"output:\\n %s\", buf.String())\n-\t\tos.Exit(1)\n-\t} else {\n-\t\tgo func() {\n-\t\t\tif err := cmd.Wait(); err != nil {\n-\t\t\t\t// don't print error if analyzer was killed by cancel\n-\t\t\t\tif ctx.Err() != context.Canceled {\n-\t\t\t\t\tioutil.ReadAll(tee)\n-\t\t\t\t\tfmt.Println(\"server exited with error:\", err)\n-\t\t\t\t\tfmt.Printf(\"output:\\n%s\", buf.String())\n-\t\t\t\t\tfailExit()\n-\t\t\t\t}\n+\trequire.NoError(err, \"can't start server\")\n+\n+\tgo func() {\n+\t\tif err := cmd.Wait(); err != nil {\n+\t\t\t// don't print error if analyzer was killed by cancel\n+\t\t\tif ctx.Err() != context.Canceled {\n+\t\t\t\tfmt.Println(\"server exited with error:\", err)\n+\t\t\t\tfmt.Printf(\"output:\\n%s\", buf.String())\n+\t\t\t\t// T.Fail cannot be called from a goroutine\n+\t\t\t\tfailExit()\n \t\t\t}\n-\t\t}()\n-\t}\n+\t\t}\n+\t}()\n \n \treturn tee, w\n }\n \n // RunCli runs lookout subcommand (not a server)\n-func RunCli(ctx context.Context, cmd string, args ...string) io.Reader {\n+func RunCli(ctx context.Context, require *require.Assertions, cmd string, args ...string) io.Reader {\n \targs = append([]string{cmd}, args...)\n \n \tvar out bytes.Buffer\n-\treviewCmd := exec.CommandContext(ctx, lookoutBin, args...)\n-\treviewCmd.Stdout = \u0026out\n-\treviewCmd.Stderr = \u0026out\n-\n-\terr := reviewCmd.Run()\n-\tif err != nil {\n-\t\tfmt.Println(\"review command returned error\")\n-\t\tfmt.Println(err)\n-\t\tfmt.Printf(\"output:\\n %s\", out.String())\n-\t\tfailExit()\n-\t}\n+\tcliCmd := exec.CommandContext(ctx, lookoutBin, args...)\n+\tcliCmd.Stdout = \u0026out\n+\tcliCmd.Stderr = \u0026out\n+\n+\terr := cliCmd.Run()\n+\trequire.NoErrorf(err,\n+\t\t\"'lookout %s' command returned error. output:\\n%s\",\n+\t\tstrings.Join(args, \" \"), out.String())\n \n \treturn \u0026out\n }\n \n // ResetDB recreates database for the test\n-func ResetDB() {\n+func ResetDB(require *require.Assertions) {\n \tdb, err := sql.Open(\"postgres\", \"postgres://postgres:postgres@localhost:5432/lookout?sslmode=disable\")\n-\tif err != nil {\n-\t\tfmt.Println(\"can't connect to DB:\", err)\n-\t\tos.Exit(1)\n-\t}\n+\trequire.NoError(err, \"can't connect to DB\")\n \n \t_, err = db.Exec(\"DROP SCHEMA public CASCADE;\")\n-\tnoDBErr(err)\n+\trequire.NoError(err, \"can't execute query\")\n \t_, err = db.Exec(\"CREATE SCHEMA public;\")\n-\tnoDBErr(err)\n+\trequire.NoError(err, \"can't execute query\")\n \t_, err = db.Exec(\"GRANT ALL ON SCHEMA public TO postgres;\")\n-\tnoDBErr(err)\n+\trequire.NoError(err, \"can't execute query\")\n \t_, err = db.Exec(\"GRANT ALL ON SCHEMA public TO public;\")\n-\tnoDBErr(err)\n+\trequire.NoError(err, \"can't execute query\")\n \n \terr = exec.Command(lookoutBin, \"migrate\").Run()\n-\tif err != nil {\n-\t\tfmt.Println(\"can't migrate DB:\", err)\n-\t\tos.Exit(1)\n-\t}\n-}\n-\n-func noDBErr(err error) {\n-\tif err != nil {\n-\t\tfmt.Println(\"can't execute query\", err)\n-\t\tos.Exit(1)\n-\t}\n+\trequire.NoError(err, \"can't migrate DB\")\n }\n \n func failExit() {","level":"error","msg":"bad hunks","time":"2018-08-22T16:54:59.011697481Z"}
{"app":"lookout","error":"posting analysis failed: bad hunk format","event-id":"8e7525f1c831ee49115fd843ed6340f086a56009","event-type":2,"level":"error","msg":"event processing failed","time":"2018-08-22T16:54:59.260271798Z"}
```

### PR DESCRIPTION
**Pull request by bzz**
Fixes #165 

CI expected to fail, as it's only a test reproducing a 🐞

### CODE REVIEW

**Comment by bzz** [comment]
Bug was fixed by @smacker in #185 this just adds extra test and error message improvement.

Should be good to merge, as soon as CI is green

File: provider/github/diff_test.go [hunk 1]
```diff
@@ -138,6 +138,44 @@ func TestParseHunks(t *testing.T) {
 			{Type: lineContext, Lines: 2},
 		},
 	}}, hunks)
+
+	hunks, err = parseHunks(largeHunk)
```
Comment (smacker) [inline provider/github/diff_test.go]: could you please move it to separate test like `TestParseHunks__issue_165`?

it's really hard to understand are those numbers correct or not. With the link to the issue, a developer later can actually generate such diff manually using commit hashes and add debug info wherever necessary if this test is failing.

Comment (bzz) [reply provider/github/diff_test.go]: Addressed in 00dc38c

File: provider/github/diff_test.go [hunk 2]
```diff
@@ -140,6 +140,45 @@ func TestParseHunks(t *testing.T) {
...
+		NewStartLine: 62,
+		NewLines:     25,
+	}, &hunk{
+		OldStartLine: 97,
+		OldLines:     85,
+		NewStartLine: 92,
+		NewLines:     59,
+	}}, hunks)
+
+	read, hunk, err := parseHunk(largeHunk)
```
Comment (smacker) [inline provider/github/diff_test.go]: I'm sorry. We don't use `parseHunk` function anymore. I forgot to remove it in my PR. Could you please remove it?

**Comment by bzz** [comment]
Drone failed on cloning

**Pull request merged**